File: internal/anchor/anchor.go
// Package anchor creates and resolves content-based line anchors.
package anchor

import (
	"errors"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

const (
	letters  = "abcdefghjkmnpqrstvwxyz"
	alphabet = "0123456789" + letters
)

var (
	refPattern        = regexp.MustCompile(`^(?:(\d+)(?:-(\d+))?#)?([a-hjkmnp-tv-z][0-9a-hjkmnp-tv-z]{3})$`)
	errInvalidRef     = errors.New("invalid anchor reference")
	errContentChanged = errors.New("content changed")
	errMultipleLines  = errors.New("multiple matching lines")
)

// Ref identifies a line or line range by position and start-line hash.
type Ref struct {
	Line int
	End  int
	Hash string
}

// Resolution reports the resolved line range and any movement from its hint.
type Resolution struct {
	Start int
	End   int
	Moved bool
	From  int
}

// Of returns the four-character anchor hash for line.
func Of(line string) string {
	const (
		basis = uint32(2166136261)
		prime = uint32(16777619)
	)

	h := basis
	for _, b := range []byte(strings.TrimSpace(line)) {
		h ^= uint32(b)
		h *= prime
	}
	v := h % 720896
	return string([]byte{
		letters[v>>15],
		alphabet[(v>>10)&31],
		alphabet[(v>>5)&31],
		alphabet[v&31],
	})
}

// Parse parses a bare, single-line, or ranged anchor reference.
func Parse(ref string) (Ref, error) {
	match := refPattern.FindStringSubmatch(ref)
	if match == nil {
		return Ref{}, fmt.Errorf("parse anchor %q: %w", ref, errInvalidRef)
	}

	r := Ref{Hash: match[3]}
	if match[1] == "" {
		return r, nil
	}

	line, err := strconv.Atoi(match[1])
	if err != nil {
		return Ref{}, fmt.Errorf("parse anchor %q line: %w", ref, err)
	}
	if line == 0 {
		return Ref{}, fmt.Errorf("parse anchor %q line must be positive: %w", ref, errInvalidRef)
	}
	r.Line = line
	r.End = line

	if match[2] == "" {
		return r, nil
	}
	end, err := strconv.Atoi(match[2])
	if err != nil {
		return Ref{}, fmt.Errorf("parse anchor %q range end: %w", ref, err)
	}
	if end < line {
		return Ref{}, fmt.Errorf("parse anchor %q range is reversed: %w", ref, errInvalidRef)
	}
	r.End = end
	return r, nil
}

// Format returns a single-line anchor reference.
func Format(line int, hash string) string {
	return fmt.Sprintf("%d#%s", line, hash)
}

// FormatRange returns a ranged anchor reference.
func FormatRange(start, end int, hash string) string {
	return fmt.Sprintf("%d-%d#%s", start, end, hash)
}

// Resolve locates r in lines using its hash and optional line hint.
func Resolve(r Ref, lines []string) (Resolution, error) {
	if r.Line > 0 && r.Line <= len(lines) && Of(lines[r.Line-1]) == r.Hash {
		return resolved(r, r.Line, len(lines)), nil
	}

	var candidates []int
	for i, line := range lines {
		if Of(line) == r.Hash {
			candidates = append(candidates, i+1)
		}
	}
	if len(candidates) == 0 {
		return Resolution{}, fmt.Errorf("anchor %s not found: %w", r.Hash, errContentChanged)
	}
	if r.Line == 0 {
		if len(candidates) > 1 {
			return Resolution{}, fmt.Errorf("anchor %s is ambiguous; candidates %v: %w", r.Hash, candidates, errMultipleLines)
		}
		return resolved(r, candidates[0], len(lines)), nil
	}

	nearest := candidates[0]
	nearestDistance := distance(nearest, r.Line)
	for _, candidate := range candidates[1:] {
		candidateDistance := distance(candidate, r.Line)
		if candidateDistance < nearestDistance {
			nearest = candidate
			nearestDistance = candidateDistance
		}
	}
	return resolved(r, nearest, len(lines)), nil
}

func resolved(r Ref, start, lineCount int) Resolution {
	end := start
	if r.Line > 0 {
		end = min(r.End, lineCount) + start - r.Line
		end = max(end, start)
		end = min(end, lineCount)
	}
	moved := r.Line > 0 && start != r.Line
	from := 0
	if moved {
		from = r.Line
	}
	return Resolution{Start: start, End: end, Moved: moved, From: from}
}

func distance(a, b int) int {
	if a < b {
		return b - a
	}
	return a - b
}

File: ios/CcPresentKit/Sources/CcPresentKit/Anchor.swift
/// Anchor creates and resolves content-based line anchors.
public enum Anchor {
    /// Ref identifies a line or line range by position and start-line hash.
    public struct Ref: Equatable {
        /// Line is the one-based position hint, or zero for a bare anchor.
        public var line: Int
        public var end: Int
        public var hash: String

        public init(line: Int, end: Int, hash: String) {
            self.line = line
            self.end = end
            self.hash = hash
        }
    }

    /// Resolution reports the resolved line range and any movement from its hint.
    public struct Resolution: Equatable {
        public var start: Int
        public var end: Int
        public var moved: Bool
        public var from: Int

        public init(start: Int, end: Int, moved: Bool, from: Int) {
            self.start = start
            self.end = end
            self.moved = moved
            self.from = from
        }
    }

    /// AnchorError reports malformed references and resolution failures.
    public enum AnchorError: Error {
        case malformed(reference: String, reason: String)
        case notFound(hash: String)
        case ambiguous(hash: String, candidates: [Int])
    }

    private static let letters = Array("abcdefghjkmnpqrstvwxyz".utf8)
    private static let alphabet = Array("0123456789abcdefghjkmnpqrstvwxyz".utf8)

    /// Of returns the four-character anchor hash for line.
    public static func of(_ line: String) -> String {
        var hash: UInt32 = 2_166_136_261
        for byte in trimmed(line).utf8 {
            hash ^= UInt32(byte)
            hash = hash &* 16_777_619
        }

        let value = Int(hash % 720_896)
        return String(decoding: [
            letters[value >> 15],
            alphabet[(value >> 10) & 31],
            alphabet[(value >> 5) & 31],
            alphabet[value & 31],
        ], as: Unicode.UTF8.self)
    }

    /// Parse parses a bare, single-line, or ranged anchor reference.
    public static func parse(_ ref: String) throws -> Ref {
        let parts = ref.split(separator: "#", omittingEmptySubsequences: false)
        if parts.count == 1 {
            guard validHash(ref) else {
                throw AnchorError.malformed(reference: ref, reason: "invalid syntax")
            }
            return Ref(line: 0, end: 0, hash: ref)
        }

        guard parts.count == 2 else {
            throw AnchorError.malformed(reference: ref, reason: "invalid syntax")
        }

        let hash = String(parts[1])
        guard validHash(hash) else {
            throw AnchorError.malformed(reference: ref, reason: "invalid hash")
        }

        let range = parts[0].split(separator: "-", omittingEmptySubsequences: false)
        guard range.count == 1 || range.count == 2, let line = decimal(range[0]) else {
            throw AnchorError.malformed(reference: ref, reason: "invalid line")
        }
        guard line > 0 else {
            throw AnchorError.malformed(reference: ref, reason: "line must be positive")
        }
        guard range.count == 2 else {
            return Ref(line: line, end: line, hash: hash)
        }
        guard let end = decimal(range[1]) else {
            throw AnchorError.malformed(reference: ref, reason: "invalid range end")
        }
        guard end >= line else {
            throw AnchorError.malformed(reference: ref, reason: "range is reversed")
        }
        return Ref(line: line, end: end, hash: hash)
    }

    /// Format returns a single-line anchor reference.
    public static func format(line: Int, hash: String) -> String {
        "\(line)#\(hash)"
    }

    /// FormatRange returns a ranged anchor reference.
    public static func formatRange(start: Int, end: Int, hash: String) -> String {
        "\(start)-\(end)#\(hash)"
    }

    /// Resolve locates ref in lines using its hash and optional line hint.
    public static func resolve(_ ref: Ref, lines: [String]) throws -> Resolution {
        if ref.line > 0, ref.line <= lines.count, of(lines[ref.line - 1]) == ref.hash {
            return resolved(ref, start: ref.line, lineCount: lines.count)
        }

        let candidates = lines.enumerated().compactMap { index, line in
            of(line) == ref.hash ? index + 1 : nil
        }
        guard !candidates.isEmpty else {
            throw AnchorError.notFound(hash: ref.hash)
        }
        if ref.line == 0 {
            guard candidates.count == 1 else {
                throw AnchorError.ambiguous(hash: ref.hash, candidates: candidates)
            }
            return resolved(ref, start: candidates[0], lineCount: lines.count)
        }

        var nearest = candidates[0]
        var nearestDistance = distance(nearest, ref.line)
        for candidate in candidates.dropFirst() {
            let candidateDistance = distance(candidate, ref.line)
            if candidateDistance < nearestDistance {
                nearest = candidate
                nearestDistance = candidateDistance
            }
        }
        return resolved(ref, start: nearest, lineCount: lines.count)
    }

    private static func validHash(_ hash: String) -> Bool {
        let bytes = Array(hash.utf8)
        return bytes.count == 4
            && letters.contains(bytes[0])
            && bytes.dropFirst().allSatisfy(alphabet.contains)
    }

    private static func decimal(_ value: Substring) -> Int? {
        guard !value.isEmpty, value.utf8.allSatisfy({ $0 >= 48 && $0 <= 57 }) else {
            return nil
        }
        return Int(value)
    }

    private static func trimmed(_ line: String) -> String {
        let scalars = line.unicodeScalars
        var start = scalars.startIndex
        while start != scalars.endIndex, isTrimSpace(scalars[start]) {
            scalars.formIndex(after: &start)
        }

        var end = scalars.endIndex
        while end != start {
            let previous = scalars.index(before: end)
            guard isTrimSpace(scalars[previous]) else {
                break
            }
            end = previous
        }
        return String(scalars[start ..< end])
    }

    private static func isTrimSpace(_ scalar: Unicode.Scalar) -> Bool {
        switch scalar.value {
        case 0x0009 ... 0x000D,
             0x0020,
             0x0085,
             0x00A0,
             0x1680,
             0x2000 ... 0x200A,
             0x2028,
             0x2029,
             0x202F,
             0x205F,
             0x3000:
            true
        default:
            false
        }
    }

    private static func resolved(_ ref: Ref, start: Int, lineCount: Int) -> Resolution {
        var end = start
        if ref.line > 0 {
            end = min(ref.end, lineCount) + start - ref.line
            end = max(end, start)
            end = min(end, lineCount)
        }
        let moved = ref.line > 0 && start != ref.line
        return Resolution(start: start, end: end, moved: moved, from: moved ? ref.line : 0)
    }

    private static func distance(_ first: Int, _ second: Int) -> Int {
        first < second ? second - first : first - second
    }
}
